case-studies/n-body/entt/main.cpp

```diff
@@ -85,74 +91,85 @@ void update(entt::registry& registry) {
-          registry.emplace<Velocity>(entity, Eigen::Vector3f::Zero());
-          registry.emplace<Acceleration>(entity, Eigen::Vector3f::Zero());
-          registry.emplace<Mass>(entity, 1e6f * radius * radius * radius);
-      }
-      
-      std::vector<std::chrono::duration<double>> measurements;
-
-      for (int i = 0; i < measurement_count; i++) {
-          auto start = std::chrono::high_resolution_clock::now();
-          for (int j = 0; j < ticks_per_measurement; j++) {
-              update(registry);
-          }
-          auto stop = std::chrono::high_resolution_clock::now();
-          measurements.emplace_back(stop - start);
-          std::cout << "." << std::flush;
-      }
-      
-      double mean = 0;
-      for (auto& measurement : measurements) {
-          mean += measurement.count();
-      }
-      mean /= measurement_count;
-      
-      double variance = 0;
-      for (auto& measurement : measurements) {
-          auto val = measurement.count();
-          variance += (val - mean) * (val - mean);
-      }
-      variance /= measurement_count - 1;
-      auto standard_deviation = sqrt(variance);
-      
-      std::cout << "\rBenchmark result for " << body_count << " bodies over " << ticks_per_measurement * measurement_count << " ticks:" << std::endl;
-      std::cout << "Avg. TPS: " << ticks_per_measurement / mean << " Avg. tick time (ms): " << mean << " Standard deviation (ms): " << standard_deviation << std::endl << std::endl;
-  }
+    int body_count;
+    int ticks_per_sample;
+    std::chrono::duration<double> samples[SAMPLE_COUNT]{};
+
+    Benchmark(int body_count, int simulated_ticks) : body_count(body_count) {
+        ticks_per_sample = simulated_ticks / SAMPLE_COUNT;
+    }
+
+    void run() {
+        entt::registry registry;
+
+        std::default_random_engine rnd;
+        std::uniform_real_distribution<float> radius_distribution(0, 1);
+        std::uniform_real_distribution<float> position_distribution(-5, 5);
+
+        for (int i = 0; i < body_count; i++) {
+            const auto entity = registry.create();
+            auto radius = radius_distribution(rnd);
+            registry.emplace<Position>(entity, Eigen::Vector3f(position_distribution(rnd), position_distribution(rnd),
+                                                               position_distribution(rnd)));
+            registry.emplace<Velocity>(entity, Eigen::Vector3f::Zero());
+            registry.emplace<Acceleration>(entity, Eigen::Vector3f::Zero());
+            registry.emplace<Mass>(entity, 1e6f * radius * radius * radius);
+        }
+
+        for (int i = -1; i < SAMPLE_COUNT; i++) {
+            auto start = std::chrono::high_resolution_clock::now();
+            for (int j = 0; j < ticks_per_sample; j++) {
+                update(registry);
+            }
+            auto stop = std::chrono::high_resolution_clock::now();
+            if (i >= 0) samples[i] = stop - start;
```

Comment:
is the first `i = -1` iteration a warmup-phase or why does it begin at `-1`?

Reply:
Yes, it’s a warmup-phase.